# 0. Cargar el corpus

In [1]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("Cornell-University/arxiv")

print("Path to dataset files:", path)

Path to dataset files: /kaggle/input/arxiv


Cargamos todos los datos del json.

In [2]:
import pandas as pd

df = pd.read_json('/kaggle/input/arxiv/arxiv-metadata-oai-snapshot.json', lines=True)
df.head()

id           submitter  \
0  0704.0001      Pavel Nadolsky   
1  0704.0002        Louis Theran   
2  0704.0003         Hongjun Pan   
3  0704.0004        David Callan   
4  0704.0005  Alberto Torchinsky   

                                             authors  \
0  C. Bal\'azs, E. L. Berger, P. M. Nadolsky, C.-...   
1                    Ileana Streinu and Louis Theran   
2                                        Hongjun Pan   
3                                       David Callan   
4           Wael Abu-Shammala and Alberto Torchinsky   

                                               title  \
0  Calculation of prompt diphoton production cros...   
1           Sparsity-certifying Graph Decompositions   
2  The evolution of the Earth-Moon system based o...   
3  A determinant of Stirling cycle numbers counts...   
4  From dyadic $\Lambda_{\alpha}$ to $\Lambda_{\a...   

                                  comments  \
0  37 pages, 15 figures; published version   
1    To appear in Graphs and Combinatorics   
2                      23 pages, 3 figures   
3                                 11 pages   
4                                     None   

                                 journal-ref                         doi  \
0                   Phys.Rev.D76:013009,2007  10.1103/PhysRevD.76.013009   
1                                       None                        None   
2                                       None                        None   
3                                       None                        None   
4  Illinois J. Math. 52 (2008) no.2, 681-689                        None   

          report-no       categories  \
0  ANL-HEP-PR-07-12           hep-ph   
1              None    math.CO cs.CG   
2              None   physics.gen-ph   
3              None          math.CO   
4              None  math.CA math.FA   

                                             license  \
0                                               None   
1  http://arxiv.org/licenses/nonexclusive-distrib...   
2                                               None   
3                                               None   
4                                               None   

                                            abstract  \
0    A fully differential calculation in perturba...   
1    We describe a new algorithm, the $(k,\ell)$-...   
2    The evolution of Earth-Moon system is descri...   
3    We show that a determinant of Stirling cycle...   
4    In this paper we show how to compute the $\L...   

                                            versions update_date  \
0  [{'version': 'v1', 'created': 'Mon, 2 Apr 2007...  2008-11-26   
1  [{'version': 'v1', 'created': 'Sat, 31 Mar 200...  2008-12-13   
2  [{'version': 'v1', 'created': 'Sun, 1 Apr 2007...  2008-01-13   
3  [{'version': 'v1', 'created': 'Sat, 31 Mar 200...  2007-05-23   
4  [{'version': 'v1', 'created': 'Mon, 2 Apr 2007...  2013-10-15   

                                      authors_parsed  
0  [[Balázs, C., ], [Berger, E. L., ], [Nadolsky,...  
1           [[Streinu, Ileana, ], [Theran, Louis, ]]  
2                                 [[Pan, Hongjun, ]]  
3                                [[Callan, David, ]]  
4  [[Abu-Shammala, Wael, ], [Torchinsky, Alberto, ]]

Reducimos el corpus eligiendo documentos aleatorios y eliminando el dataframe anterior.

In [3]:
df_reducido = df.sample(n=100000, random_state=42)
del df

In [4]:
import gc

gc.collect()

0

In [5]:
df_reducido.info()

<class 'pandas.core.frame.DataFrame'>
Index: 100000 entries, 635536 to 1158450
Data columns (total 14 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   id              100000 non-null  object
 1   submitter       99479 non-null   object
 2   authors         100000 non-null  object
 3   title           100000 non-null  object
 4   comments        73032 non-null   object
 5   journal-ref     31357 non-null   object
 6   doi             43770 non-null   object
 7   report-no       6504 non-null    object
 8   categories      100000 non-null  object
 9   license         84353 non-null   object
 10  abstract        100000 non-null  object
 11  versions        100000 non-null  object
 12  update_date     100000 non-null  object
 13  authors_parsed  100000 non-null  object
dtypes: object(14)
memory usage: 11.4+ MB


# 1. Preprocesamiento de Datos

In [6]:
df_reducido['id'].nunique()

100000

In [7]:
df_reducido["id"] = df_reducido["id"].astype(str)

Eliminamos columnas innecesarias.

In [8]:
df_reducido = df_reducido.drop(columns=["doi", "report-no", "categories", "license", "versions", "update_date", "authors_parsed"])

In [9]:
df_reducido.head()

id         submitter  \
635536   1506.07092  Andrei Faminskii   
2185841  2411.03413     Xinyuan Zhang   
1044621  1810.13434     Megan Ansdell   
2124626  2408.04444       Dandan Chen   
1027118  1809.07102    Kirubin Pillay   

                                                   authors  \
635536                                 Andrei V. Faminskii   
2185841  Xiaoyu Chen, Zongchen Chen, Yitong Yin, Xinyua...   
1044621  Megan Ansdell, Yani Ioannou, Hugh P. Osborn, M...   
2124626                         Dandan Chen and Zhiguo Liu   
1027118                  Kirubin Pillay and Maarten De Vos   

                                                     title  \
635536   An initial-boundary value problem for three-di...   
2185841           Rapid Mixing at the Uniqueness Threshold   
1044621  Scientific Domain Knowledge Improves Exoplanet...   
2124626   New proofs of theorems on q-orthogonal functions   
1027118  A unifying Bayesian approach for preterm brain...   

                                                  comments journal-ref  \
635536   23 pages. arXiv admin note: text overlap with ...        None   
2185841  Remove the incorrectly claimed square-root spe...        None   
1044621        8 pages, 5 figures, accepted to ApJ Letters        None   
2124626                                               None        None   
1027118  5 pages, 2 figures. Submitted for Neural Infor...        None   

                                                  abstract  
635536     An initial-boundary value problem with homog...  
2185841  Over the past decades, a fascinating computati...  
1044621    Space-based missions such as Kepler, and soo...  
2124626    In this paper, we establish a $q$-integral f...  
1027118    Preterm newborns undergo various stresses th...

In [10]:
df_reducido.info()

<class 'pandas.core.frame.DataFrame'>
Index: 100000 entries, 635536 to 1158450
Data columns (total 7 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   id           100000 non-null  object
 1   submitter    99479 non-null   object
 2   authors      100000 non-null  object
 3   title        100000 non-null  object
 4   comments     73032 non-null   object
 5   journal-ref  31357 non-null   object
 6   abstract     100000 non-null  object
dtypes: object(7)
memory usage: 6.1+ MB


Ahora viene lo bueno, la información está en la columna "abstract" así que limpiaremos el texto.

In [11]:
import pandas as pd
import re
import nltk
from nltk.corpus import stopwords

nltk.download('stopwords')
stop_words = set(stopwords.words('english'))

[nltk_data] Downloading package stopwords to /usr/share/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [12]:
def limpiar_texto(text):
    text = text.lower()
    
    text = re.sub(r'\n', ' ', text)
    text = re.sub(r'[^a-z0-9\s]', '', text)
    
    text = re.sub(r'\s+', ' ', text).strip()
    
    words = text.split()
    words = [w for w in words if w not in stop_words]
    
    return " ".join(words)

In [13]:
df_reducido['abstract_clean'] = df_reducido['abstract'].apply(limpiar_texto)

In [14]:
print(df_reducido[['abstract', 'abstract_clean']].head())

                                                  abstract  \
635536     An initial-boundary value problem with homog...   
2185841  Over the past decades, a fascinating computati...   
1044621    Space-based missions such as Kepler, and soo...   
2124626    In this paper, we establish a $q$-integral f...   
1027118    Preterm newborns undergo various stresses th...   

                                            abstract_clean  
635536   initialboundary value problem homogeneous diri...  
2185841  past decades fascinating computational phase t...  
1044621  spacebased missions kepler soon tess provide l...  
2124626  paper establish qintegral formula using orthog...  
1027118  preterm newborns undergo various stresses may ...  


# 2. Representación mediante embeddings

In [15]:
from sentence_transformers import SentenceTransformer
import numpy as np

# 1. Cargar el modelo SBERT
model = SentenceTransformer('all-MiniLM-L6-v2')

2026-01-28 17:42:54.059940: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1769622174.300478      55 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1769622174.369360      55 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1769622174.934853      55 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1769622174.934884      55 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1769622174.934888      55 computation_placer.cc:177] computation placer alr

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [16]:
sentences = df_reducido['abstract_clean'].tolist()

embeddings = model.encode(sentences, batch_size=64, show_progress_bar=True)

Batches:   0%|          | 0/1563 [00:00<?, ?it/s]

# 3. Recuperación inicial

In [20]:
!pip install faiss-cpu

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.8/23.8 MB 80.8 MB/s eta 0:00:00:00:0100:01


In [21]:
import faiss

dimension = embeddings.shape[1]
index = faiss.IndexFlatL2(dimension)

embeddings_faiss = np.array(embeddings).astype('float32')

# Agregar vectores al índice
index.add(embeddings_faiss)

print(f"Índice FAISS listo con {index.ntotal} vectores.")

Índice FAISS listo con 100000 vectores.


In [24]:
def buscar_similares(query, k=5):
    query_vector = model.encode([query]).astype('float32')
    
    distancias, indices = index.search(query_vector, k)
    
    return df_reducido.iloc[indices[0]]

Recuperamos información con una query.

In [25]:
resultados = buscar_similares("Advancements in large language models and transformers", k=3)
print(resultados[['title', 'abstract']])

                                                     title  \
1972602         A mathematical perspective on Transformers   
1303800  On the Computational Power of Transformers and...   
1113418  Dynamic Evaluation of Transformer Language Models   

                                                  abstract  
1972602  Transformers play a central role in the inner ...  
1303800    Transformers are being used extensively acro...  
1113418    This research note combines two methods that...  


# 4. Re-ranking de resultados

In [26]:
from sentence_transformers import CrossEncoder

reranker_model = CrossEncoder('cross-encoder/ms-marco-MiniLM-L-6-v2')

config.json:   0%|          | 0.00/794 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/132 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

In [29]:
def buscar_con_reranking(query, top_k_faiss=50, top_n_final=5):
    query_vector = model.encode([query]).astype('float32')
    distancias, indices = index.search(query_vector, top_k_faiss)
    
    candidatos = df_reducido.iloc[indices[0]].copy()
    
    # Reranking con Cross-Encoder
    pares = [[query, doc] for doc in candidatos['abstract_clean']]
    
    scores = reranker_model.predict(pares)
    
    candidatos['rerank_score'] = scores
    resultado_final = candidatos.sort_values(by='rerank_score', ascending=False)
    
    return resultado_final.head(top_n_final)

In [30]:
query_usuario = "Advancements in large language models and transformers"
resultados_optimizados = buscar_con_reranking(query_usuario)

print(resultados_optimizados[['title', 'rerank_score']])

                                                     title  rerank_score
2178690  FIRP: Faster LLM inference via future intermed...      4.644126
1935477  Towards Understanding How Transformers Learn I...      4.455449
1972602         A mathematical perspective on Transformers      4.349786
1916478  A Paradigm Shift in Machine Translation: Boost...      4.278616
2336554  Is Random Attention Sufficient for Sequence Mo...      3.526567


# 6. Evaluación del sistema

Ya que el corpus no contiene qrels, nos lo inventamos nosotros.

In [31]:
qrels = {
    "large language models": ["123.456", "789.012"],
    "quantum computing": ["456.789"],
    "dark matter observations": ["111.222", "333.444"]
}

Definimos una función de cálculo de métricas para FAISS y Cross Encoder.

In [34]:
def calcular_metricas(query, ids_relevantes, k=10):
    # Recuperación con FAISS
    query_vec = model.encode([query]).astype('float32')
    _, idx_faiss = index.search(query_vec, k)
    recuperados_faiss = df_reducido.iloc[idx_faiss[0]]['id'].tolist()
    
    # Recuperación con Reranker
    res_rerank = buscar_con_reranking(query, top_k_faiss=k, top_n_final=k)
    recuperados_rerank = res_rerank['id'].tolist()
    
    def evaluar(recuperados, relevantes):
        hits = [1 if r in relevantes else 0 for r in recuperados]
        p_at_k = sum(hits) / len(hits)
        
        mrr = 0
        for i, h in enumerate(hits):
            if h == 1:
                mrr = 1 / (i + 1)
                break
        return p_at_k, mrr

    met_faiss = evaluar(recuperados_faiss, ids_relevantes)
    met_rerank = evaluar(recuperados_rerank, ids_relevantes)
    
    return met_faiss, met_rerank

In [35]:
# Ejecutar evaluación
for query, ids in qrels.items():
    f, r = calcular_metricas(query, ids)
    print(f"Query: {query}")
    print(f"  FAISS -> P@10: {f[0]:.2f}, MRR: {f[1]:.2f}")
    print(f"  Rerank -> P@10: {r[0]:.2f}, MRR: {r[1]:.2f}")

Query: large language models
  FAISS -> P@10: 0.00, MRR: 0.00
  Rerank -> P@10: 0.00, MRR: 0.00
Query: quantum computing
  FAISS -> P@10: 0.00, MRR: 0.00
  Rerank -> P@10: 0.00, MRR: 0.00
Query: dark matter observations
  FAISS -> P@10: 0.00, MRR: 0.00
  Rerank -> P@10: 0.00, MRR: 0.00


# 7. Análisis de resultados

Los resultados entre FAISS y Cross Encoder son muy diferentes, ya que los documentos recuperados son reordenados de manera diferente:

title  \
1972602         A mathematical perspective on Transformers   
1303800  On the Computational Power of Transformers and...   
1113418  Dynamic Evaluation of Transformer Language Models

title  rerank_score
2178690  FIRP: Faster LLM inference via future intermed...      4.644126
1935477  Towards Understanding How Transformers Learn I...      4.455449
1972602         A mathematical perspective on Transformers      4.349786
1916478  A Paradigm Shift in Machine Translation: Boost...      4.278616
2336554  Is Random Attention Sufficient for Sequence Mo...      3.526567

Con qrels bien definidos, podríamos entender mejores resultados para este corpus y el ranking de estos dos modelos. 

Lo mismo pasa al intentar evaluar el modelo, en el que nos inventamos algunas queries con los documentos relevantes.